Src/Support/Google.Apis.Auth.Tests/OAuth2/TokenRefreshManagerTests.cs

```diff
@@ -0,0 +1,283 @@
+                for (int i = 0; i < concurrentRefreshCount; i++)
+                {
+                    tokenTasks[i] = trm.GetAccessTokenForRequestAsync(CancellationToken.None);
+                }
+                delayTask.SetResult(0);
+                var tokens = await Task.WhenAll(tokenTasks);
+
+                // It's non-deterministic if the refresh will return the previous or just-retreived token
+                var expectedAccessTokenA = $"AccessToken{iteration}";
+                var expectedAccessTokenB = $"AccessToken{iteration - 1}";
+                foreach (var token in tokens)
+                {
+                    Assert.True(token == expectedAccessTokenA || token == expectedAccessTokenB);
+                }
+
+                clock.UtcNow += TimeSpan.FromSeconds(TokenResponse.TokenRefreshTimeWindowSeconds - TokenResponse.TokenHardExpiryTimeWindowSeconds);
+            }
+        }
+
+        [Fact]
+        public async Task ExpiredToken()
+        {
+            var clock = new MockClock { UtcNow = new DateTime(2010, 1, 1, 0, 0, 0, DateTimeKind.Utc) };
+            var logger = new NullLogger();
+            int refreshFnCount = 0;
+            string accessToken = null;
+            TokenRefreshManager trm = null;
+            trm = new TokenRefreshManager(RefreshFn, clock, logger);
+            Task<bool> RefreshFn(CancellationToken ct)
+            {
+                trm.Token = new TokenResponse
+                {
+                    AccessToken = accessToken,
+                    ExpiresInSeconds = 60 * 60,
+                    IssuedUtc = clock.UtcNow
+                };
+                Interlocked.Increment(ref refreshFnCount);
+                return Task.FromResult(true);
+            }
+
+            accessToken = "AccessToken1";
+            var accessToken1 = await trm.GetAccessTokenForRequestAsync(CancellationToken.None);
+            clock.UtcNow += TimeSpan.FromHours(24);
+            accessToken = "AccessToken2";
+            var accessToken2 = await trm.GetAccessTokenForRequestAsync(CancellationToken.None);
+
+            Assert.Equal(2, Interlocked.Add(ref refreshFnCount, 0));
+            Assert.Equal("AccessToken1", accessToken1);
+            Assert.Equal("AccessToken2", accessToken2);
+        }
+
+        [Fact]
+        public async Task CancelDuringRefresh()
+        {
+            var clock = new MockClock { UtcNow = new DateTime(2010, 1, 1, 0, 0, 0, DateTimeKind.Utc) };
+            var logger = new NullLogger();
+            int refreshCalled = 0;
+            int refreshCompleted = 0;
+            TokenRefreshManager trm = null;
+            trm = new TokenRefreshManager(RefreshFn, clock, logger);
+            async Task<bool> RefreshFn(CancellationToken ct)
+            {
+                Interlocked.Increment(ref refreshCalled);
+                await Task.Delay(TimeSpan.FromSeconds(60), ct);
+                // Will never get here
```

Comment:
Will we not? I'd expect that we would, but after the test has completed. The cancellation token received by the refresh function won't be cancelled by anything, will it?

Reply:
Changed so the test really never does get here.